lib/MSSA/MemRegion.cpp

```diff
@@ -557,3 +557,83 @@ void MRGenerator::modRefAnalysis(PTACallGraphNode* callGraphNode, WorkList& work
         }
     }
 }
+
+/*!
+ * Determine whether a CallSite instruction can mod or ref
+ * any memory location
+ */
+ModRefInfo MRGenerator::getModRefInfo(CallSite cs) {
+    bool ref = hasRefSideEffectOfCallSite(cs);
+    bool mod = hasModSideEffectOfCallSite(cs);
+
+    if (mod && ref)
+        return ModRefInfo::ModRef;
+    else if (ref)
+        return ModRefInfo::Ref;
+    else if (mod)
+        return ModRefInfo::Mod;
+    else
+        return ModRefInfo::NoModRef;
+}
+
+/*!
+ * Determine whether a CallSite instruction can mod or ref
+ * a specific memory location pointed by V
+ */
+ModRefInfo MRGenerator::getModRefInfo(CallSite cs, const Value* V) {
+    bool ref = false;
+    bool mod = false;
+
+    if (pta->getPAG()->hasValueNode(V)) {
+        PointsTo pts(pta->getPts(pta->getPAG()->getValueNode(V)));
+        if (hasRefSideEffectOfCallSite(cs) && getRefSideEffectOfCallSite(cs).intersects(pts))
+            ref = true;
+        if (hasModSideEffectOfCallSite(cs) && getModSideEffectOfCallSite(cs).intersects(pts))
+            mod = true;
+    }
+
+    if (mod && ref)
+        return ModRefInfo::ModRef;
+    else if (ref)
+        return ModRefInfo::Ref;
+    else if (mod)
+        return ModRefInfo::Mod;
+    else
+        return ModRefInfo::NoModRef;
+}
+
+/*!
+ * Determine mod-ref relations between two CallSite instructions
+ */
+ModRefInfo MRGenerator::getModRefInfo(CallSite cs1, CallSite cs2) {
+    bool ref = false;
+    bool mod = false;
+
+    /// return NoModRef neither two callsites ref or mod any memory
+    if (getModRefInfo(cs1) == ModRefInfo::NoModRef || getModRefInfo(cs2) == ModRefInfo::NoModRef)
+        return ModRefInfo::NoModRef;
+
+    PointsTo cs1Ref = getRefSideEffectOfCallSite(cs1);
```

Comment:
If you are not modifying the Ref and Mod sets, you can consider using references to avoid copy construction.

`PointsTo cs1Ref = getRefSideEffectOfCallSite(cs1)`
or
`const PointsTo cs1Ref = getRefSideEffectOfCallSite(cs1)'